bin/faucet/src/client.rs

```diff
@@ -0,0 +1,382 @@
+
+        let delta = executed_tx.account_delta().clone();
+
+        let proven_transaction =
+            transaction_prover.prove_transaction(executed_tx).map_err(|err| {
+                FaucetError::InternalServerError(format!("Failed to prove transaction: {}", err))
+            })?;
+
+        let request = SubmitProvenTransactionRequest {
+            transaction: proven_transaction.to_bytes(),
+        };
+
+        self.rpc_api
+            .submit_proven_transaction(request)
+            .await
+            .map_err(|err| FaucetError::InternalServerError(err.to_string()))?;
+
+        self.data_store.update_faucet_account(&delta).map_err(|err| {
+            FaucetError::InternalServerError(format!("Failed to update account: {}", err))
+        })?;
+
+        Ok(())
+    }
+
+    pub fn get_faucet_id(&self) -> AccountId {
+        self.id
+    }
+}
+
+#[derive(Clone)]
+pub struct FaucetDataStore {
+    faucet_account: Rc<RefCell<Account>>,
+    seed: Word,
+    block_header: BlockHeader,
+    chain_mmr: ChainMmr,
+}
+
+impl FaucetDataStore {
+    pub fn new(
+        faucet_account: Rc<RefCell<Account>>,
+        seed: Word,
+        root_block_header: BlockHeader,
+        root_chain_mmr: ChainMmr,
+    ) -> Self {
+        Self {
+            faucet_account,
+            seed,
+            block_header: root_block_header,
+            chain_mmr: root_chain_mmr,
+        }
+    }
+
+    fn update_faucet_account(&mut self, delta: &AccountDelta) -> Result<(), FaucetError> {
+        self.faucet_account
+            .borrow_mut()
+            .apply_delta(delta)
+            .map_err(|err| FaucetError::InternalServerError(err.to_string()))
+    }
+}
+
+impl DataStore for FaucetDataStore {
+    fn get_transaction_inputs(
+        &self,
+        account_id: AccountId,
+        _block_ref: u32,
+        _notes: &[NoteId],
+    ) -> Result<TransactionInputs, DataStoreError> {
+        let account = self.faucet_account.borrow();
+        if account_id != account.id() {
+            return Err(DataStoreError::AccountNotFound(account_id));
+        }
+
+        let empty_input_notes =
+            InputNotes::new(Vec::new()).map_err(DataStoreError::InvalidTransactionInput)?;
+
+        TransactionInputs::new(
+            account.clone(),
+            account.is_new().then_some(self.seed),
+            self.block_header,
+            self.chain_mmr.clone(),
+            empty_input_notes,
+        )
+        .map_err(DataStoreError::InvalidTransactionInput)
+    }
+
+    fn get_account_code(&self, account_id: AccountId) -> Result<ModuleAst, DataStoreError> {
+        let account = self.faucet_account.borrow();
+        if account_id != account.id() {
+            return Err(DataStoreError::AccountNotFound(account_id));
+        }
+
+        let module_ast = account.code().module().clone();
+        Ok(module_ast)
+    }
+}
+
+struct FaucetAuthenticator {
+    secret_key: AuthSecretKey,
+    rng: RefCell<ChaCha20Rng>,
```

Comment:
as for what was said above, what happens if we wrap this with a `Mutex`/ `RwLock` instead of a `RefCell`?

Reply:
[It was suggested](https://github.com/0xPolygonMiden/miden-base/pull/714#pullrequestreview-2087440701) to remove this struct in favor of the `BasicAuthenticator` which currently also uses a `RefCell`. [In the future](https://github.com/0xPolygonMiden/miden-client/issues/358#issuecomment-2135305518) this may change and we could remove the `unsafe impl Send+Sync`.